### ISSUE TITLE
Clean up the usage of schema lock

### ISSUE THREAD

**Issue by mbautin** [yugabyte/yugabyte-db#249]
It looks like we currently don't take the schema lock on reads/writes. See if that is a problem during concurrent workload + alter table.

**Comment by rthallamko3** [comment]
@amitanandaiyer , Is this relevant?

**Comment by kriteshranjan** [comment]
Closing this issue based on AI validation. cc: @m-iancu , @VamsidharGunturu

**Comment by kriteshranjan** [comment]
AI validation notes (`yb-issue-249-analysis.md`):

# Analysis: yugabyte/yugabyte-db#249 (no Jira cross-link)

Issue: "Clean up the usage of schema lock" — kind/enhancement, filed 2018-04-27
by mbautin, still OPEN. Body: "It looks like we currently don't take the schema
lock on reads/writes. See if that is a problem during concurrent workload +
alter table."

Reviewed: 2026-08-15 on master `b89bc62335` (b89bc6233565f83487971bcbef4bc5b8532a42f3, 2026-08-14).

## Verdict
FIXED (app verdict: Fixed) — the requested cleanup was delivered by commit
`42a9d18035` (2020-07-10, "[#4904] Remove the schema_lock_", Phorge D8882),
and the "is concurrent workload + alter a problem?" question was answered by
explicit replacement mechanisms that are present on current master. Nothing
remains to do; the issue can be closed as completed (superseded by CLOSED
issues #4039 and #4904).

## Root cause / current state
Filed-era state (SHA `0d07add220f5`, 2018-04-27): the tablet held a Kudu-
inherited `mutable rw_semaphore schema_lock_` (`src/yb/tablet/tablet.h:490`)
whose comment claimed "Alter schema must run when no reads/writes are in
progress", taken exclusively in the ALTER prepare path
(`src/yb/tablet/tablet.cc:1035` `operation_state->AcquireSchemaLock(&schema_lock_)`,
DCHECK at `tablet.cc:1058`; holder in
`src/yb/tablet/operations/alter_schema_operation.cc:74-82`). No read or write
path acquired it — the shared-mode write-side acquisition
(`WriteOperationState::AcquireSchemaLock`, `shared_lock<rw_semaphore>`) had
been deleted by `17277e65fa` (2017-11-29, "Remove old columnar table type"),
orphaning the exclusive side. mbautin's premise was therefore correct: the
lock guarded nothing.

Resolution timeline:
- `e65bda06cc` (2020-06-28, "[#4039] Use ScopedRWOperationPause during YCQL
  schema change to pause write operations") — introduced explicit
  alter-vs-write mutual exclusion.
- `42a9d18035` (2020-07-10, "[#4904] Remove the schema_lock_") — deleted the
  vestigial lock; commit message: "we rely on the RWOperationCounters and the
  Row locks to enforce mutual exclusion." Author: Amitanand Aiyer — the same
  engineer pinged (unanswered) in the issue's lone 2023 comment.
- #4039 ("Schema verification should be done upon Apply") and #4904 ("Get rid
  of the schema lock altogether") are both CLOSED; #4904's body explicitly
  frames itself as the follow-up that replaces `schema_lock_`, noting the
  write path stopped using it at `17277e6`.

Current master (`b89bc62335`) safety mechanisms replacing the lock:
1. Alter-vs-write mutual exclusion: `TabletServiceAdminImpl::AlterSchema`
   pauses write permits before applying a schema change —
   `src/yb/tserver/tablet_service.cc:1171-1186`
   (`pause_writes = tablet.tablet->PauseWritePermits(...)` at `:1178`;
   YCQL gated by `FLAGS_disable_alter_vs_write_mutual_exclusion`, YSQL
   unconditional; skipped only for CDCSDK retention-barrier-only alters).
   `Tablet::PauseWritePermits` = `src/yb/tablet/tablet.cc:4481-4486`, over
   `write_ops_being_submitted_counter_` (`src/yb/tablet/tablet.h:1351`,
   named "Tablet schema" at `tablet.cc:748`). Every write acquires the permit
   at `src/yb/tablet/write_query.cc:1089` (`tablet->GetPermitToWrite(deadline())`).
2. Per-request schema-version validation on reads AND writes: YCQL read
   `src/yb/tablet/tablet.cc:2289-2293` and PGSQL read `tablet.cc:2440-2442`
   return `*_SCHEMA_VERSION_MISMATCH` for stale-schema requests; write-side
   checks at `src/yb/tablet/write_query.cc:1239`, `:1266` (YCQL) and `:1569`,
   `:1596` (PGSQL).
The only `schema_lock_` remaining in the tree is an unrelated `rw_spinlock`
protecting the WAL-header schema copy (`src/yb/consensus/log.h:606`,
`log.cc:1859`, `log.cc:2198`).

## Trigger matrix (behavior-delta: predicted claim vs observed evidence)
| # | claim | evidence command | observed |
|---|-------|------------------|----------|
| 1 | 2018 tree: tablet `schema_lock_` exists, ALTER-only exclusive | `git grep -n 'schema_lock\|AcquireSchemaLock' 0d07add220f5 -- src/yb` | CONFIRMED — tablet.h:490, tablet.cc:1035/1058, alter_schema_operation.cc:74-82 |
| 2 | 2018 tree: no read/write path takes it (issue premise) | same grep | CONFIRMED — no other users besides unrelated consensus/log.cc |
| 3 | Writes took it shared-mode before 2017-11-29 | `git show 17277e65fa \| grep AcquireSchemaLock` | CONFIRMED — removes `WriteOperationState::AcquireSchemaLock` (`shared_lock<rw_semaphore>`) |
| 4 | Master: lock removed (cleanup delivered) | `git grep -n schema_lock_ -- src/yb/tablet` | CONFIRMED — no matches; removal commit 42a9d18035 |
| 5 | Master: alter pauses writes explicitly | grep `PauseWritePermits` tablet_service.cc | CONFIRMED — :1171-1186; write permit at write_query.cc:1089 |
| 6 | Master: stale-schema reads/writes rejected per request | grep `SCHEMA_VERSION_MISMATCH` src/yb/tablet | CONFIRMED — tablet.cc:2289-2293/2440-2442; write_query.cc:1239/1266/1569/1596 |

## Reproduction
N/A — enhancement/cleanup request; no crash signature. Validation is the
source-level behavior delta above; all evidence commands are rerunnable from
any yugabyte-db checkout (see matrix). No containers or builds required.

## Versions tested
Source-level comparison only: filed-era SHA `0d07add220f5` (2018-04-27) vs
master `b89bc62335` (2026-08-14). Master affected: NO — the vestigial lock no
longer exists; replacement mechanisms verified at the citations above.

## Proposed fix direction
No code change needed. Recommended action: close #249 as completed,
referencing `42a9d18035`/D8882 (#4904) as the delivering commit and #4039 for
the alter-vs-write exclusion half of the question. (Posting a closing comment
requires user authorization — not granted this session.)

## Test strategy
N/A (no fix to test). Existing coverage of the replacement mechanisms:
alter-vs-write exclusion is exercised by YCQL/YSQL DDL-concurrency tests
around `FLAGS_disable_alter_vs_write_mutual_exclusion` and
`FLAGS_TEST_skip_aborting_active_transactions_during_schema_change`
(`src/yb/tserver/tablet_service.cc:1174,1191`); schema-version mismatch paths
are exercised wherever `SCHEMA_VERSION_MISMATCH` retries are tested.

## Confidentiality check
Clean — contains only public GitHub issue text, public commit hashes/messages,
and upstream source citations. No customer data, PII, secrets, or internal
references.
